dlib/invoke.h

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2016  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_INVOKE_Hh_
+#define DLIB_INVOKE_Hh_
+
+#ifndef DLIB_CPLUSPLUS
+    #if defined(_MSVC_LANG ) && !defined(__clang__)
+        #define DLIB_CPLUSPLUS  (_MSC_VER == 1900 ? 201103L : _MSVC_LANG )
+    #else
+        #define DLIB_CPLUSPLUS  __cplusplus
+    #endif
+#endif
+
+#if defined(_MSC_VER)
+    #if _MSC_VER >= 1900
+        #define DLIB_HAVE_INVOKE 1
+    #endif
+#else
+    #if DLIB_CPLUSPLUS >= 201703L
+        #define DLIB_HAVE_INVOKE 1
+    #endif
+#endif
```

Comment:
Having backports of new C++ stuff is cool.  And at some point we should switch it to just use the one in the standard library.  But at any given time we should commit to one implementation so we don't have to have this kind of stuff.  Like just don't have any `#ifdef` stuff.  Make it always use the implementation you wrote.  Then someday we can delete it an make it just say `using std::whatever`.

But in the mean time it's better to avoid needing to fight with visual studio's nonsense.  And like, I'm not convinced this is robust since `_MSC_VER == 1900 ? 201103L : _MSVC_LANG` ought to be at least `_MSC_VER <= 1900 ? 201103L : _MSVC_LANG`.

Trying to weave through the parts of MSVC support for current C++ standards is always awful.  It would be a miracle if we didn't get people commenting later about how in some particular update of MSVC or with some setting something in here was failing.  MSVC is the main reason I wait so long to let dlib use current C++ features :(


Anyway, can you remove the #ifdef stuff and just have the one version of what's here?  Same for the other file.